plugins/filter/to_ini.py

```diff
@@ -0,0 +1,105 @@
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2023, Steffen Scheib <steffen@scheib.me>
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from __future__ import absolute_import, division, print_function
+
+DOCUMENTATION = r'''
+  name: to_ini
+  short_description: Converts a dictionary to the INI file format
+  version_added: 8.2.0
+  author: Steffen Scheib (@sscheib)
+  description:
+    - Converts a dictionary to the INI file format.
+  options:
+    _input:
+      description: The dictionary that should be converted to the INI format.
+      type: dictionary
+      required: true
+'''
+
+EXAMPLES = r'''
+  - name: Define a dictionary
+    ansible.builtin.set_fact:
+      my_dict:
+        section_name:
+          key_name: 'key value'
+
+        another_section:
+          connection: 'ssh'
+
+  - name: Write dictionary to INI file
+    ansible.builtin.copy:
+      dest: /tmp/test.ini
+      content: '{{ my_dict | community.general.to_ini }}'
+
+  # /tmp/test.ini will look like this:
+  # [section_name]
+  # key_name = key value
+  #
+  # [another_section]
+  # connection = ssh
+'''
+
+RETURN = r'''
+  _value:
+    description: A string formatted as INI file.
+    type: string
+'''
+
+
+__metaclass__ = type
+
+from ansible.errors import AnsibleFilterError
+from ansible.module_utils.common._collections_compat import MutableMapping
+from ansible.module_utils.six.moves import StringIO
+from ansible.module_utils.six.moves.configparser import ConfigParser
+from ansible.module_utils.common.text.converters import to_native
+
+
+class IniParser(ConfigParser):
+    ''' Implements a configparser which sets the correct optionxform '''
+
+    def __init__(self):
+        super().__init__()
+        self.optionxform = str
+
+
+def to_ini(obj):
+    ''' Read the given dict and return an INI formatted string '''
+
+    if not isinstance(obj, MutableMapping):
```

Comment:
Out of curiosity: why `MutableMapping` and not simply `Mapping`?

Reply:
Essentially, I saw it on other modules being used, but I cannot recall exactly where. Other modules in `community.general` use it as well (e.g.: https://github.com/ansible-collections/community.general/blob/main/plugins/callback/log_plays.py#L75C37-L75C37, https://github.com/ansible-collections/community.general/blob/main/plugins/inventory/proxmox.py#L326C56-L326C56, https://github.com/ansible-collections/community.general/blob/main/plugins/inventory/virtualbox.py#L99C17-L99C17).
If you'd like me to implement the more generic `Mapping`, I can do, no problem :)

Reply:
`log_plays` actually needs it to be mutable; for the other two I guess that someone copied it from somewhere else without thinking about this aspect... I think it would be better to use `Mapping` here instead of `MutableMapping`, since there's no need for the mapping to be mutable. While it's not possible to create an immutable mapping in ansible-core with what's included (to my knowledge), you could create a lookup, filter plugin, or something else that produces an immutable mapping and feed that into this filter.